[JSC] Add aligned label annotation to offlineasm

eugeneia opened the pull request:
#### 71f803e6ee62ac5bcf8449a8e9cb1d4fd601f74a
<pre>
[JSC] Add aligned label annotation to offlineasm
<a href="https://bugs.webkit.org/show_bug.cgi?id=270699">https://bugs.webkit.org/show_bug.cgi?id=270699</a>

Reviewed by Yusuke Suzuki and Justin Michaud.

To express some programs like the InPlaceInterpreter we need to align
instructions to specific boundaries (such as 256).

We currently to do this via `emit &quot;.balign 256, 0xcc&quot;&apos; (for instance)
but this is not reliable on armv7 as its globaladdr implementation can
insert code unpredictably (from the point of view of the offlineasm
user), hence breaking the alignment.

This adds an aligned label annotation which implements this
functionality in offlineasm.

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/offlineasm/asm.rb:
* Source/JavaScriptCore/offlineasm/ast.rb:
* Source/JavaScriptCore/offlineasm/backends.rb:
* Source/JavaScriptCore/offlineasm/parser.rb:
* Source/JavaScriptCore/offlineasm/transform.rb:

Canonical link: <a href="https://commits.webkit.org/276085@main">https://commits.webkit.org/276085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/80e767de49617fc794cf5da181661202586b711f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43073 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45474 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45697 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39197 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25895 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19521 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35615 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43647 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19199 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37103 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16609 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38163 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1128 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/36530 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/39322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47235 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42699 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17983 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14779 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42401 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19525 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41054 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9721 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19703 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49709 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19157 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10044 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
```diff
@@ -587,13 +587,23 @@ JSValue CLoop::execute(OpcodeID entryOpcodeID, void* executableAddress, VM* vm,
+#define OFFLINE_ASM_ALIGN_TRAP(align) ".balign " #align ", 0xcc\n"
+#elif CPU(ARM)
+#define OFFLINE_ASM_ALIGN_TRAP(align) ".balignw " #align ", 0xde00\n"
+#endif
```
When moving this here, deleted the comments explaining the magic numbers, which are apparently `brk`, `int 3`, and `udf`. I think it would be a good idea to retain those helpful comments in some form.

eugeneia replied:
Good catch, I agree! Should be addressed in https://github.com/WebKit/WebKit/commit/80e767de49617fc794cf5da181661202586b711f

justinmichaud reviewed (approved):
r=me with Darin's feedback addressed

darinadler commented on Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
```diff
@@ -4302,7 +4293,7 @@ unimplementedInstruction(_simd_i64x2_extmul_high_i32x4_u)
 unimplementedInstruction(_simd_f32x4_abs)
 unimplementedInstruction(_simd_f32x4_neg)
-reservedOpcode(0xFDC201)
+reservedOpcode(0xFDE201)
```
This change seems mysterious. Change log doesn’t mention it. Is the reason for it obvious to people who have more expertise than I do?

kmiller68 replied:
I think it's a copy paste error or typo. I don't think it caused problems before because prior to this change the passed value to `reservedOpcode` was ignored. Now `reservedOpcode` adds a global label so there would be a conflict with the previous `reservedOpcode(0xFDC201)` a few lines above.

eugeneia replied:
@kmiller68 is correct, that's the reason. The new opcode should be the correct one to go in place.

Constellation reviewed (approved):
r=me too

eugeneia committed:
[JSC] Add aligned label annotation to offlineasm
https://bugs.webkit.org/show_bug.cgi?id=270699

Reviewed by Yusuke Suzuki and Justin Michaud.

To express some programs like the InPlaceInterpreter we need to align
instructions to specific boundaries (such as 256).

We currently to do this via `emit ".balign 256, 0xcc"' (for instance)
but this is not reliable on armv7 as its globaladdr implementation can
insert code unpredictably (from the point of view of the offlineasm
user), hence breaking the alignment.

This adds an aligned label annotation which implements this
functionality in offlineasm.

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/LowLevelInterpreter.cpp:
* Source/JavaScriptCore/offlineasm/asm.rb:
* Source/JavaScriptCore/offlineasm/ast.rb:
* Source/JavaScriptCore/offlineasm/backends.rb:
* Source/JavaScriptCore/offlineasm/parser.rb:
* Source/JavaScriptCore/offlineasm/transform.rb:

Canonical link: https://commits.webkit.org/276085@main
